Userland/Libraries/LibCore/FileLikeIODevice.cpp

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2021, Ali Mohammad Pur <ali.mpfard@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/ByteBuffer.h>
+#include <AK/PrintfImplementation.h>
+#include <LibCore/FileLikeIODevice.h>
+#include <LibCore/Notifier.h>
+#include <sys/select.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+namespace Core {
+
+FileLikeIODevice::FileLikeIODevice(Object* parent)
+    : IODevice(parent)
+{
+}
+
+FileLikeIODevice::~FileLikeIODevice()
+{
+}
+
+const char* FileLikeIODevice::error_string() const
+{
+    return strerror(m_error);
+}
+
+size_t FileLikeIODevice::read(Bytes bytes)
+{
+    if (m_fd < 0) {
+        set_fatal_error();
+        return 0;
+    }
+
+    if (bytes.is_empty())
+        return 0;
+
+    auto buffer_ptr = bytes.data();
+    size_t remaining_buffer_space = bytes.size();
+
+    auto nread = ::read(m_fd, buffer_ptr, remaining_buffer_space);
+    if (nread < 0) {
+        set_error(errno);
+        return 0;
+    }
+
+    if (nread == 0) {
+        set_eof(true);
+        return 0;
+    }
+
+    return static_cast<size_t>(nread);
+}
+
+bool FileLikeIODevice::discard_or_error(size_t count)
+{
+    // FIXME: This can probably be implemented in a less wasteful way...
+    auto buffer = ByteBuffer::create_uninitialized(count);
+    if (read(buffer.bytes()) < count) {
+        set_recoverable_error();
+        return false;
+    }
+    return true;
+}
+
+ByteBuffer FileLikeIODevice::read(size_t max_size)
+{
+    auto buffer = ByteBuffer::create_uninitialized(max_size);
+    buffer.trim(read(buffer.bytes()));
+    return buffer;
+}
+
+bool FileLikeIODevice::can_read_from_fd() const
+{
+    // FIXME: Can we somehow remove this once Core::Socket is implemented using non-blocking sockets?
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(m_fd, &rfds);
+    struct timeval timeout {
+        0, 0
+    };
+
+    for (;;) {
+        if (select(m_fd + 1, &rfds, nullptr, nullptr, &timeout) < 0) {
+            if (errno == EINTR)
+                continue;
+            perror("IODevice::can_read_from_fd: select");
+            return false;
+        }
+        break;
+    }
+    return FD_ISSET(m_fd, &rfds);
+}
+
+bool FileLikeIODevice::can_read() const
+{
+    return can_read_from_fd();
+}
+
+ByteBuffer FileLikeIODevice::read_all()
+{
+    off_t file_size = 0;
+    struct stat st;
+    int rc = fstat(fd(), &st);
+    if (rc == 0)
+        file_size = st.st_size;
+
+    Vector<u8> data;
+    data.ensure_capacity(file_size);
+
+    while (true) {
+        char read_buffer[4096];
+        int nread = ::read(m_fd, read_buffer, sizeof(read_buffer));
```

Comment:
`auto` would be more appropriate here, as `::read` returns `ssize_t`, not `int`.

Reply:
I have a feeling that this came from existing code, but yeah, `auto` makes more sense.